src/core/public/context/context.ts

```diff
@@ -0,0 +1,172 @@
+   * @returns A Promise for the new context object.
+   */
+  createContext(
+    plugin: PluginName,
+    baseContext: Partial<TContext>,
+    ...contextArgs: TProviderParameters
+  ): Promise<TContext>;
+}
+
+/** @internal */
+export class ContextContainerImplementation<
+  TContext extends {},
+  TProviderParameters extends any[] = []
+> implements ContextContainer<TContext, TProviderParameters> {
+  /**
+   * Used to map contexts to their providers and associated plugin. In registration order which is tightly coupled to
+   * plugin load order.
+   */
+  private readonly contextProviders = new Map<
+    keyof TContext,
+    {
+      provider: ContextProvider<TContext, keyof TContext, TProviderParameters>;
+      plugin?: PluginName;
+    }
+  >();
+  /** Used to keep track of which plugins registered which contexts for dependency resolution. */
+  private readonly contextNamesByPlugin = new Map<PluginName, Array<keyof TContext>>();
+
+  /**
+   * @param pluginDependencies - A map of plugins to an array of their dependencies.
+   */
+  constructor(private readonly pluginDependencies: ReadonlyMap<PluginName, PluginName[]>) {}
+
+  public register<TContextName extends keyof TContext>(
+    contextName: TContextName,
+    provider: ContextProvider<TContext, TContextName, TProviderParameters>,
+    plugin?: PluginName
+  ): this {
+    if (this.contextProviders.has(contextName)) {
+      throw new Error(`Context provider for ${contextName} has already been registered.`);
+    }
+
+    this.contextProviders.set(contextName, { provider, plugin });
+
+    if (plugin) {
+      this.contextNamesByPlugin.set(plugin, [
+        ...(this.contextNamesByPlugin.get(plugin) || []),
+        contextName,
+      ]);
+    }
+
+    return this;
+  }
+
+  public async createContext(
+    pluginName: PluginName,
+    baseContext: Partial<TContext> = {},
+    ...contextArgs: TProviderParameters
+  ): Promise<TContext> {
+    const contextNamesForPlugin = (plug: PluginName): Array<keyof TContext> => {
+      const pluginDeps = this.pluginDependencies.get(plug)!;
+      return flatten(pluginDeps.map(p => this.contextNamesByPlugin.get(p)!));
+    };
+
+    // Set of all contexts depended on by this plugin
+    const neededContexts = new Set([
+      ...(this.contextNamesByPlugin.get(pluginName) || []),
+      ...contextNamesForPlugin(pluginName),
+    ]);
+
+    return [...this.contextProviders]
+      .filter(
+        ([contextName, { plugin }]) =>
+          // Contexts we depend on OR provided by core
+          neededContexts.has(contextName) || plugin === undefined
+      )
+      .reduce(
+        async (contextPromise, [contextName, { provider, plugin }]) => {
+          const resolvedContext = await contextPromise;
+
+          // If the provider is not from a plugin, give access to the entire
+          // context built so far (this is only possible for providers registered
+          // by the service owner).
```

Comment:
This will have to be made impossible by the service owner (which can be a Core service or a plugin).

The service owner that exposes this context container's register method should populate the pluginName argument based on the scoped service contract pattern in #40822